### PR TITLE
rsx: PS3 Native frame limiter improvements, add Infinite frame limiter

### PR DESCRIPTION
**Pull request by elad335**
* Do not wait on DEVICE 0x30 semaphore, it seems like it is something to do with queue command synchronization.
 - This also fixes cellGcmSetFlipWithWaitLabel which is built specifically to enable accurate RSX flipping time, its waiting command is confirmed to be placed **AFTER** DEVICE 0x30 waiting.
* Fix default VSYNC state to be enabled. (and set it to be enabled in cellGcmSetVBlankFrequency and cellVideOutConfigure as well)
* Add experimental "Infinite" frame limiter mode.
* Fix spurious enabling of second vblank.

### CODE REVIEW

**Comment by elad335** [comment]
I've added a new frame limiter called "Infinite" which basically imposes no limits whatsoever and sends an additional VBLANK each frame 
This is an extreme "Off" frame limiter type.

**Comment by kd-11** [comment]
> I've added a new frame limiter called "Infinite" which basically imposes no limits whatsoever and sends an additional VBLANK each frame . This is an extreme "Off" frame limiter type. I've also added "Vblank" type which is the same behavior as before #12045.

We already know games are using vblank count for timing. This makes things worse imo.
The correct solution here is to not wait for the flip semaphore but to implement hardware triple buffering for RSX. A display queue is used and we check the semaphore during flip()

**Comment by kd-11** [comment]
If the semaphore is not set and we have a frame to consume, go ahead and consume it. Hopefully by the time that frame is ready the old one will have been cleared

**Comment by kd-11** [comment]
The display queue already exists by the way, it just needs some tweaks to bind it to a RSX semaphore. I'd like to extend this in future by using the native GPU's semaphore system which does a very similar thing so its important we get the architecture right.

**Comment by solarmystic** [comment]
With Framelimit set to VBLANK, performance is back to normal again. Frame times are also smoother. 

image

**Comment by Asinin3** [comment]
God of War 3 still has more spikes and slower than before when using auto framelimiter.

Before HW Framelimiter:
13626_neozcull_9EEgKAK67A

After:
13627_hw_framelimit_77xzUCiHis

After this fixup PR:
framelimiter_g9TJhHnWTh

**Comment by elad335** [comment]
I've switched the auto and vblank's demeanor, also please test the new "infinite" frame limiter which is supposed to be a replacement for upping the vblank rate for higher framerates in many games.

**Comment by Asinin3** [comment]
Vblank limiter is worse now.

old commit: 
framelimiter_YA32mtm87j
New Commit:
hw_tsEd1a16WN

Infinite in new commit:
hw_2Yl0TzH8HZ
1-3fps worse than Off for me.

**Comment by elad335** [comment]
For clarification - off is the fastest option, infinite is supposed to be used with games that have an internal vblank-based frame limiter.

**Comment by solarmystic** [comment]
With these latest commits, God of War Ascension behaves as follows with each framelimiter option:-

Auto - frame time graph is not so spikey but performance is still lower compared to pre HW limiter behaviour
Auto FL

Vblank -  the spikiest frametimes, and the least performant option.
VBLANK FL

Infinite - better than Vblank and Auto, and matches pre HW limiter behaviour in that scene.
Infinite FL

Pre hardware HW limiter build (0.0.22-13626) for comparison:-
rpcs3_2022_05_21_17_52_50_543

File: rpcs3/Emu/RSX/rsx_methods.cpp
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_methods.cpp:110]: This is obviously inverted logic here. This is a frame acquire semaphore, its supposed to insert a barrier here that states "do not display the payload until this value changes".
Now, obviously this whole thing is simulated, but this is the wrong semantic here. This write is a release operation, not an acquire.

File: rpcs3/Emu/system_config.h
Comment (kd-11) [inline rpcs3/Emu/system_config.h:159]: I don't see why this exists and why it is an option.

File: rpcs3/Emu/RSX/RSXThread.cpp
Comment (kd-11) [inline rpcs3/Emu/RSX/RSXThread.cpp:3280]: This feels like it should just be part of "infinite" vblank mode and does not need the extra loop option

File: rpcs3/Emu/RSX/RSXThread.cpp
Comment (kd-11) [inline rpcs3/Emu/RSX/RSXThread.cpp:179]: This is silent log spam. If it is unexpected, raise it to error and leave a note about notying developers. That should get us faster feedback if this functionality is misunderstood.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp]: Removed.

File: rpcs3/rpcs3qt/emu_settings.cpp
Comment (Megamouse) [inline rpcs3/rpcs3qt/emu_settings.cpp:958]:
```suggestion
		case frame_limit_type::infinite: return tr("Infinite", "Frame limit");
```

File: rpcs3/rpcs3qt/tooltips.h
Comment (Megamouse) [inline rpcs3/rpcs3qt/tooltips.h:161]: why?

Comment (elad335) [reply rpcs3/rpcs3qt/tooltips.h]: Missed the update to tooltips.

File: rpcs3/rpcs3qt/tooltips.h
Comment (Megamouse) [inline rpcs3/rpcs3qt/tooltips.h:161]: fps limitless sounds weird.
maybe call it uncapped framerate or something like that

Comment (elad335) [reply rpcs3/rpcs3qt/tooltips.h:161]: uncapped is already off, limitless emphasizes the additional removal of vblank induced limit imo.

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: Where is the counterpart to disable this? We know PS3 had engines that had tearing when fps dropped below a threshold (e.g cryengine 2)

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: I see sometimes in logs that the games reset it afterward and after cellVideoOutConfigure, I will add to it as well.

File: rpcs3/Emu/RSX/rsx_methods.cpp
Comment (kd-11) [reply rpcs3/Emu/RSX/rsx_methods.cpp:114]: This is still a problem.

Comment (elad335) [reply rpcs3/Emu/RSX/rsx_methods.cpp]: Moved it to semaphore_release.

File: rpcs3/Emu/RSX/RSXThread.cpp
Comment (kd-11) [inline rpcs3/Emu/RSX/RSXThread.cpp:3262]: If I'm reading this correctly the logic here is essentially "do not flip until a fresh vblank signal is received". You have the vblank signal as the flip release and this is an acquire operation here. Two problems:
1. There is still the correct event sent to us which you are now short-circuiting in semaphore-acquire. I'm not sure why this approach is to be considered better as it is essentially a hackier approach. Note that this "PS3 native" sync is to be used with games that have trouble with the normal PC-friendly framelimiter system, so extra bells and whistles are unneeded for this path. Is it that it doesn't work properly with HLE? I'm scratching my head on why I should prefer this version over the existing one which more closely follows what the PS3 does.
2. This whole set of variables and checks is not correctly labeled to match how display systems are architected with clients acquiring the display and the server releasing it in a cycle. Not a blocker by itself, but its not ideal and I would end up having to change it.

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp
Comment (kd-11) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: Hmm, it is unlikely that they are calling videoOutConfigure during gameplay as that is a heavy operation. It is more likely to be toggled using this method. Is it known what happens if a5 is 0?

Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: I'm assuming that games also react to the unused CellVideoOutCallback. (maybe it's called when the display is changed)

Comment (kd-11) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: It's controlled by CellGcmSetFlipMode. You can choose 3 modes there, 1 for vsync, 1 for scanline sync and the last mode for all out anarchy. We only need to set vsync to engage when the first mode is chosen.

File: rpcs3/Emu/RSX/RSXThread.cpp
Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: I'm quite confident DEVICE 0x30 is something to do with the queue command, as it's inserted by gcm for all display sync options and occurs before semaphore wait by cellGcmSetFlipWithWaitLabel.
This function exists specifically to allow accurate timing of flipping and this is a usage example:
image
Here is label 0 is used, notice that it resides after DEVICE 0x30 waiting 
Furthermore, flips are executed along with the vblank signal, not after, I guess that's why TLoU regressed with the original pr.

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:552]: > I'm assuming that games also react to the unused CellVideoOutCallback. (maybe it's called when the display is changed)

I've checked and no game uses cellVideoOutRegisterCallback.

File: rpcs3/Emu/RSX/RSXThread.cpp
Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: From the trace above, we can see there are 2 signals:
The only difference is that there are 2 cond vars that I see, one to ack the enqueue and another to wait for vertical sync.
Server:
```
	label_0x30.wait_signal();
	handle_enqueued_head();
	label_0x30.reset()
```
Client:
```
	enqueue();
	label_0x30.signal();
	label_0x0.wait_signal();
	flip();
```
My guess here is that 0x0 is signaled by the hardware vblank system, 0x30 just acknowledges that a head has been enqueued. So, your PR addresses this by having the system use vblank instead of enqueue signal.
While this is fine, the second critique still holds. The implementation is convoluted and not clear at all.
My suggestion to solve this once and for all is to have a special kind of cond var with waiting for known values. Simply replace my rushed solution with the wait_for_flip_sema with proper condvar.wait() semantics to make the flow clearer.

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: For simplicity I have omitted the 0x10 logic, I have no idea where that signal is for.
For 0x30 handling, we have the frame enqueue logic in sys_rsx and rsx_thread::end_frame() for that.
For 0x0 we can signal from vblank thread.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: No I have HLEd the game this is a cellGcmSetFlipWithWaitLabel call with 0 address. It doesn't appear in other gsmes which uses other flip functions.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: Whis is named _cellGcmSetFlipCommandWithWaitLabel on firmware.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: Here is an LLE flip without waiting on label.
image

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: So for "normal" flip the 0x0 var is absent? That simplifies the design then and makes the 0x30_ack a wait for vblank.
Again, my biggest problem with the way it is now is that the "documentation is the code" philosophy doesn't hold up because the design is all over the place. This isn't a new problem, but if it is now well understood, we have to do it properly and this PR already modifies the whole thing anyway. I only understand what is going on because I'm familiar with most of the code there.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: 0x30 has nothing to do with vblank, it appears in all sync modes even with VSYNC turned off and besides the whole purpose of _cellGcmSetFlipCommandWithWaitLabel is to time the flip very closely to what CELL decides. So if 0x30 is a vblank semaphore it defies all logic as to why the function exists in the first place.

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:3262]: It seems my point is getting lost in translation, that doesn't matter it's just an ack, whether its from vblank or not. I'm not happy with the implementation, but if we're being honest, you're unlikely to restructure it, so I'll merge and change it later when I have time.

**Review by kd-11 — APPROVED** [review]

**Pull request merged**
